### ISSUE TITLE
Helper functions for non-time series

### ISSUE THREAD

**Issue by daviddiviny** [MattCowgill/readabs#74]
Probably not a priority for this package, but putting this code here if it is helpful in thinking about a way to download "data cubes" from abs.gov.au. It needs a bit more abstraction for it to work as part of a package. 

```r
library(tidyverse)
library(rvest)
library(glue)

releases_url <- "https://www.abs.gov.au/AUSSTATS/abs@.nsf/second+level+view?ReadForm&prodno=6160.0.55.001&viewtitle=Weekly%20Payroll%20Jobs%20and%20Wages%20in%20Australia&&tabname=Past%20Future%20Issues&prodno=6160.0.55.001"
releases_page <- read_html(releases_url) 
releases_table <- tibble(release = releases_page %>%  html_nodes("#mainpane a") %>% html_text(),
                         url_suffix = releases_page %>%  html_nodes("#mainpane a") %>% html_attr("href"))
#get the page for the latest data
latest_text <- releases_table %>% 
  filter(grepl("(Latest)", release)) %>% 
  pull(release) %>% 
  str_remove("Weekly Payroll Jobs and Wages in Australia ") %>% 
  str_remove(" \\(Latest\\)") %>% 
  str_replace_all(" ", "%20")

latest_url <- glue("https://www.abs.gov.au/AUSSTATS/abs@.nsf/DetailsPage/6160.0.55.001{latest_text}?OpenDocument")

latest_page <- read_html(latest_url)
#Find the url for the download
download_url_suffix <- tibble(url = latest_page %>% html_nodes("a") %>% html_attr("href")) %>% 
  filter(grepl("do004", url)) %>% 
  pull(url)

url <- glue("https://www.abs.gov.au{download_url_suffix}")

#==================download data======================
download.file(url = url,
              destfile = "raw-data/6160055001_do004.xlsx", 
              mode = "wb")


```

**Comment by daviddiviny** [comment]
Hi @MattCowgill  - I've made this code more general (e.g. it now works with EQ09 Datacube from the LFS.  Let me know if you are interested in this becoming part of `readabs` otherwise we will add to an internal package. 

```r 
library(tidyverse)
series <- "6291.0.55.003"
cube_string <- "eq09"
path  <- "raw-data"
latest <- TRUE


if(latest == FALSE & is.null(date)) {stop("latest is false and date is NULL. Please supply a value for date.")}
  

releases_url <- glue::glue("https://www.abs.gov.au/AUSSTATS/abs@.nsf/second+level+view?ReadForm&prodno={series}&&tabname=Past%20Future%20Issues")

releases_page <- xml2::read_html(releases_url) 

releases_table <- tibble::tibble(release = releases_page %>%  rvest::html_nodes("#mainpane a") %>% rvest::html_text(),
                         url_suffix = releases_page %>%  rvest::html_nodes("#mainpane a") %>% rvest::html_attr("href"))

#get the page for the latest data
if(latest == TRUE){
date <- releases_table %>% 
  dplyr::filter(grepl("(Latest)", release)) %>% 
  dplyr::pull(release) %>% 
  stringr::str_remove(" \\(Latest\\)") %>% 
  stringr::str_extract("Week ending \\d+\\s{1}\\w+ \\d+$|(Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:tember)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?) \\d+$|\\d+$") %>% 
  stringr::str_replace_all(" ", "%20")
}

#find download page
download_url <- glue::glue("https://www.abs.gov.au/AUSSTATS/abs@.nsf/DetailsPage/{series}{date}?OpenDocument")
download_page <- xml2::read_html(download_url)


#Find the url for the download
download_url_suffix <- tibble::tibble(url = download_page %>% rvest::html_nodes("a") %>% rvest::html_attr("href")) %>% 
  dplyr::filter(grepl(tolower(cube_string), url)) %>% 
  dplyr::slice(1) %>% #this gets the first result which is typically the xlsx file rather than the zip
  dplyr::pull(url) 

#==================download file======================
file_url <- glue::glue("https://www.abs.gov.au{download_url_suffix}")

download_object <- httr::GET(file_url)

filename <- basename(download_object$url)
filepath <- file.path(path, filename)

writeBin(httr::content(download_object, "raw"), filepath)

```

**Comment by MattCowgill** [comment]
Hi @daviddiviny this is great, sorry it's taken me a while to get back to you. I'd be happy to incorporate it into the package -- maybe with an 'experimental' tag?

I had a teleconference with the ABS last week and they told me that the new website will have persistent URLs for the latest release of datacubes. This might make this process easier

**Comment by daviddiviny** [comment]
Great, will get to the documentation when I get a chance. Did the ABS say anything about API access for datacubes? 

This line is probably the dodgiest so good to test out some other examples. It works for the Weekly Payrolls and Labour Force so that is a start. 
```r
download_url <- glue::glue("https://www.abs.gov.au/AUSSTATS/abs@.nsf/DetailsPage/{series}{date}?OpenDocument")

```

**Comment by MattCowgill** [comment]
Great, will test more broadly. Will you do a PR? 

Their plan is:
- Improve the ABS.Stat API (which includes some, but far from all data that is in datacubes)
- Maintain and improve the Time Series Directory (which is what `readabs` uses for time series)
- Use persistent URLs to fetch the latest release of any table

As far as I could gather, they do not have plans to add more data to ABS.Stat, which is the real problem there.

**Comment by daviddiviny** [comment]
Thanks @MattCowgill  - when do you intend to put this on CRAN?

**Comment by MattCowgill** [comment]
Hi @daviddiviny -- I intend to submit an update this week. It's currently failing tests - I think the problem is the legacy `read_abs_sdmx()` function, which may have to be deprecated. I'll figure it out this week and submit to CRAN, will let you know when I have. FYI @zmeers I think the SDMX function may have to go away

**Comment by MattCowgill** [comment]
FYI @daviddiviny I just tried to submit the update to CRAN and received the following error: 

```
Error: Submission failed:
The submission team is on vacation from Aug 14, 2020 to Aug 24, 2020.
During this time the submission of packages is not possible.
Sorry for any inconvenience. 
```

I'll try again next week

**Comment by daviddiviny** [comment]
Hey @MattCowgill  - are you able to have another go with CRAN rather than being on Twitter?

**Comment by MattCowgill** [comment]
Hi @daviddiviny I resubmitted it yesterday. It has failed due to an inability to build the vignette, which I cannot reproduce locally nor on Travis. I will try to puzzle through it.

**Comment by daviddiviny** [comment]
Great, thanks! Sorry for trolling then!

**Comment by MattCowgill** [comment]
The new version is up on CRAN now. Binaries not yet available, but source is there. https://cran.r-project.org/web/packages/readabs/index.html

**Comment by daviddiviny** [comment]
Any idea why it is not showing 0.4.3 on CRAN yet @MattCowgill ?

**Comment by MattCowgill** [comment]
Not sure what you mean @daviddiviny... Version 0.4.3.2 is there on CRAN, with all binaries built. Same version as here on Github. The `download_abs_data_cube()` function is there... https://cran.r-project.org/web/packages/readabs/index.html

**Comment by daviddiviny** [comment]
Doh. For some reason I thought we were waiting on 0.4.3.
